### PR TITLE
[2064]: Add stochastic sampler (sample_dpmpp_sde)

### PR DESCRIPTION
**Pull request by nipunjindal**
Discussion: #2064 

Implements Stochastic Sampler based on the original k-diffusion implementation by [Katherine Crowson](https://github.com/crowsonkb/k-diffusion/blob/41b4cb6df0506694a7776af31349acf082bf6091/k_diffusion/sampling.py#L543).

**DPM + Stochastic solver + karras**
```
import diffusers
import torch
from diffusers import DPMSolverSDEScheduler

seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = DPMSolverSDEScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

display(image)
```

diffusers-sde

**k-diffusion + sample_dpmpp_sde + karras**
FYI: There is a bug in the kdiffusion code where BrownianTreeNoiseSampler is not being passed the seed, so the run for sample_dpmp_sde gives back different results each time, fixed after giving the seed though. So make sure you do that change before comparing.

```
import torch
from diffusers import StableDiffusionKDiffusionPipeline

seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4"
).to("cuda")
pipe.set_scheduler("sample_dpmpp_sde")

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(
    prompt, generator=generator, num_inference_steps=20, use_karras_sigmas=True
).images[0]

display(image)
```

kdiffusion-sde

Tensor Comparison

**DIffusers** 

```
Final unet input  tensor(999., device='cuda:0', dtype=torch.float64) tensor(0.1752, device='cuda:0')
Final unet input  tensor(980.6143, device='cuda:0', dtype=torch.float64) tensor(0.6369, device='cuda:0')
Final unet input  tensor(961.7328, device='cuda:0', dtype=torch.float64) tensor(0.0530, device='cuda:0')
Final unet input  tensor(941.6854, device='cuda:0', dtype=torch.float64) tensor(0.6597, device='cuda:0')
Final unet input  tensor(921.0421, device='cuda:0', dtype=torch.float64) tensor(0.3369, device='cuda:0')
Final unet input  tensor(899.0386, device='cuda:0', dtype=torch.float64) tensor(0.6140, device='cuda:0')
Final unet input  tensor(876.3138, device='cuda:0', dtype=torch.float64) tensor(1.1630, device='cuda:0')
Final unet input  tensor(851.9903, device='cuda:0', dtype=torch.float64) tensor(0.8114, device='cuda:0')
Final unet input  tensor(826.7906, device='cuda:0', dtype=torch.float64) tensor(1.1928, device='cuda:0')
Final unet input  tensor(799.7012, device='cuda:0', dtype=torch.float64) tensor(1.6681, device='cuda:0')
Final unet input  tensor(771.5520, device='cuda:0', dtype=torch.float64) tensor(1.3307, device='cuda:0')
Final unet input  tensor(741.1702, device='cuda:0', dtype=torch.float64) tensor(1.9652, device='cuda:0')
Final unet input  tensor(709.5297, device='cuda:0', dtype=torch.float64) tensor(2.3679, device='cuda:0')
Final unet input  tensor(675.2842, device='cuda:0', dtype=torch.float64) tensor(2.6295, device='cuda:0')
Final unet input  tensor(639.6147, device='cuda:0', dtype=torch.float64) tensor(2.7159, device='cuda:0')
Final unet input  tensor(601.0142, device='cuda:0', dtype=torch.float64) tensor(1.9187, device='cuda:0')
Final unet input  tensor(560.9754, device='cuda:0', dtype=torch.float64) tensor(2.3180, device='cuda:0')
Final unet input  tensor(517.9016, device='cuda:0', dtype=torch.float64) tensor(2.1788, device='cuda:0')
Final unet input  tensor(473.7489, device='cuda:0', dtype=torch.float64) tensor(1.2311, device='cuda:0')
Final unet input  tensor(426.9748, device='cuda:0', dtype=torch.float64) tensor(1.1715, device='cuda:0')
Final unet input  tensor(380.1317, device='cuda:0', dtype=torch.float64) tensor(1.0301, device='cuda:0')
Final unet input  tensor(331.8835, device='cuda:0', dtype=torch.float64) tensor(0.8350, device='cuda:0')
Final unet input  tensor(285.3175, device='cuda:0', dtype=torch.float64) tensor(0.6734, device='cuda:0')
Final unet input  tensor(239.3123, device='cuda:0', dtype=torch.float64) tensor(0.2582, device='cuda:0')
Final unet input  tensor(197.0806, device='cuda:0', dtype=torch.float64) tensor(0.0601, device='cuda:0')
Final unet input  tensor(157.5366, device='cuda:0', dtype=torch.float64) tensor(0.0512, device='cuda:0')
Final unet input  tensor(123.3951, device='cuda:0', dtype=torch.float64) tensor(0.4463, device='cuda:0')
Final unet input  tensor(93.3974, device='cuda:0', dtype=torch.float64) tensor(0.7651, device='cuda:0')
Final unet input  tensor(69.2569, device='cuda:0', dtype=torch.float64) tensor(1.0655, device='cuda:0')
Final unet input  tensor(49.4825, device='cuda:0', dtype=torch.float64) tensor(0.9873, device='cuda:0')
Final unet input  tensor(34.6974, device='cuda:0', dtype=torch.float64) tensor(1.1490, device='cuda:0')
Final unet input  tensor(23.3862, device='cuda:0', dtype=torch.float64) tensor(1.0358, device='cuda:0')
Final unet input  tensor(15.4816, device='cuda:0', dtype=torch.float64) tensor(1.0499, device='cuda:0')
Final unet input  tensor(9.7733, device='cuda:0', dtype=torch.float64) tensor(1.0784, device='cuda:0')
Final unet input  tensor(5.9918, device='cuda:0', dtype=torch.float64) tensor(1.0480, device='cuda:0')
Final unet input  tensor(3.4160, device='cuda:0', dtype=torch.float64) tensor(1.0344, device='cuda:0')
Final unet input  tensor(1.7830, device='cuda:0', dtype=torch.float64) tensor(1.0791, device='cuda:0')
Final unet input  tensor(0.7298, device='cuda:0', dtype=torch.float64) tensor(1.0777, device='cuda:0')
Final unet input  tensor(0., device='cuda:0', dtype=torch.float64) tensor(1.0990, device='cuda:0')
```

**K-Diffusion**

```
Final unet input  tensor([998.9999, 998.9999], device='cuda:0') tensor(0.1752, device='cuda:0')
Final unet input  tensor([980.6143, 980.6143], device='cuda:0') tensor(0.6377, device='cuda:0')
Final unet input  tensor([961.7327, 961.7327], device='cuda:0') tensor(0.0528, device='cuda:0')
Final unet input  tensor([941.6853, 941.6853], device='cuda:0') tensor(0.6598, device='cuda:0')
Final unet input  tensor([921.0420, 921.0420], device='cuda:0') tensor(0.3360, device='cuda:0')
Final unet input  tensor([899.0386, 899.0386], device='cuda:0') tensor(0.6142, device='cuda:0')
Final unet input  tensor([876.3136, 876.3136], device='cuda:0') tensor(1.1638, device='cuda:0')
Final unet input  tensor([851.9903, 851.9903], device='cuda:0') tensor(0.8107, device='cuda:0')
Final unet input  tensor([826.7906, 826.7906], device='cuda:0') tensor(1.1929, device='cuda:0')
Final unet input  tensor([799.7012, 799.7012], device='cuda:0') tensor(1.6687, device='cuda:0')
Final unet input  tensor([771.5519, 771.5519], device='cuda:0') tensor(1.3305, device='cuda:0')
Final unet input  tensor([741.1702, 741.1702], device='cuda:0') tensor(1.9648, device='cuda:0')
Final unet input  tensor([709.5297, 709.5297], device='cuda:0') tensor(2.3665, device='cuda:0')
Final unet input  tensor([675.2842, 675.2842], device='cuda:0') tensor(2.6266, device='cuda:0')
Final unet input  tensor([639.6146, 639.6146], device='cuda:0') tensor(2.7142, device='cuda:0')
Final unet input  tensor([601.0141, 601.0141], device='cuda:0') tensor(1.9176, device='cuda:0')
Final unet input  tensor([560.9752, 560.9752], device='cuda:0') tensor(2.3170, device='cuda:0')
Final unet input  tensor([517.9015, 517.9015], device='cuda:0') tensor(2.1783, device='cuda:0')
Final unet input  tensor([473.7487, 473.7487], device='cuda:0') tensor(1.2312, device='cuda:0')
Final unet input  tensor([426.9746, 426.9746], device='cuda:0') tensor(1.1714, device='cuda:0')
Final unet input  tensor([380.1315, 380.1315], device='cuda:0') tensor(1.0316, device='cuda:0')
Final unet input  tensor([331.8833, 331.8833], device='cuda:0') tensor(0.8341, device='cuda:0')
Final unet input  tensor([285.3175, 285.3175], device='cuda:0') tensor(0.6740, device='cuda:0')
Final unet input  tensor([239.3123, 239.3123], device='cuda:0') tensor(0.2576, device='cuda:0')
Final unet input  tensor([197.0806, 197.0806], device='cuda:0') tensor(0.0589, device='cuda:0')
Final unet input  tensor([157.5366, 157.5366], device='cuda:0') tensor(0.0499, device='cuda:0')
Final unet input  tensor([123.3950, 123.3950], device='cuda:0') tensor(0.4455, device='cuda:0')
Final unet input  tensor([93.3974, 93.3974], device='cuda:0') tensor(0.7638, device='cuda:0')
Final unet input  tensor([69.2569, 69.2569], device='cuda:0') tensor(1.0639, device='cuda:0')
Final unet input  tensor([49.4825, 49.4825], device='cuda:0') tensor(0.9856, device='cuda:0')
Final unet input  tensor([34.6973, 34.6973], device='cuda:0') tensor(1.1472, device='cuda:0')
Final unet input  tensor([23.3862, 23.3862], device='cuda:0') tensor(1.0342, device='cuda:0')
Final unet input  tensor([15.4816, 15.4816], device='cuda:0') tensor(1.0487, device='cuda:0')
Final unet input  tensor([9.7733, 9.7733], device='cuda:0') tensor(1.0771, device='cuda:0')
Final unet input  tensor([5.9918, 5.9918], device='cuda:0') tensor(1.0473, device='cuda:0')
Final unet input  tensor([3.4160, 3.4160], device='cuda:0') tensor(1.0339, device='cuda:0')
Final unet input  tensor([1.7829, 1.7829], device='cuda:0') tensor(1.0782, device='cuda:0')
Final unet input  tensor([0.7297, 0.7297], device='cuda:0') tensor(1.0768, device='cuda:0')
Final unet input  tensor([0., 0.], device='cuda:0') tensor(1.0979, device='cuda:0')
```

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_dpmsolver_sde.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torchsde
```
Comment (nipunjindal) [inline src/diffusers/schedulers/scheduling_dpmsolver_sde.py:20]: As `torchsde` may not be a hard dependency, a possible alternative could be to pass the BrownianTree noise sampler from the outside into the step function, and document this approach in the function's documentation. Do you have any other suggestions on how to proceed?

**Comment by nipunjindal** [comment]
Hello @patrickvonplaten @sayakpaul,

Could you kindly provide your feedback on the pull request? I will add the necessary tests once I receive general approval on the direction of the changes.
Thanks much!

File: src/diffusers/schedulers/scheduling_dpmsolver_sde.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_dpmsolver_sde.py:20]: I think we can have `torchsde` as a soft dependency like we have `kdiffusion`. 

@patrickvonplaten WDYT?

File: src/diffusers/schedulers/scheduling_dpmsolver_sde.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+    Args:
+        num_train_timesteps (`int`): number of diffusion steps used to train the model. beta_start (`float`): the
+        starting `beta` value of inference. beta_end (`float`): the final `beta` value. beta_schedule (`str`):
+            the beta schedule, a mapping from a beta range to a sequence of betas for stepping the model. Choose from
+            `linear` or `scaled_linear`.
+        trained_betas (`np.ndarray`, optional):
+            option to pass an array of betas directly to the constructor to bypass `beta_start`, `beta_end` etc.
+            options to clip the variance used when adding noise to the denoised sample. Choose from `fixed_small`,
+            `fixed_small_log`, `fixed_large`, `fixed_large_log`, `learned` or `learned_range`.
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_dpmsolver_sde.py:127]: Where is this coming from? Seems to be merged with `trained_betas`.

**Review by sayakpaul — COMMENTED** [review]
I think this ship has sailed pretty nicely already. 

I am okay with having `torchsde` as a soft dependency rather than passing an external `BrownianTree` sampler as I think that complicates the user workflow here.

**Comment by mkhennoussi** [comment]
Thanks for this amazing work @nipunjindal ! I've tried the sampler and the results looks great ! (Doesn't work with img2img pipeline, maybe this is only on my side).

File: src/diffusers/schedulers/scheduling_dpmsolver_sde.py (hunk 2 above)
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_dpmsolver_sde.py]: Fixed, this shouldn't be here. heun_discrete had the same issue, so fixed that as well.

**Comment by nipunjindal** [comment]
@mkhennoussi Can you share the repro steps? This is what I tried.

```
from io import BytesIO

import requests
import torch
from diffusers import DPMSolverSDEScheduler, StableDiffusionImg2ImgPipeline
from PIL import Image

# load the pipeline
device = "cuda"
pipe = StableDiffusionImg2ImgPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4"
).to(device)
pipe.scheduler = DPMSolverSDEScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

# let's download an initial image
url = "https://raw.githubusercontent.com/CompVis/stable-diffusion/main/assets/stable-samples/img2img/sketch-mountains-input.jpg"

response = requests.get(url)
init_image = Image.open(BytesIO(response.content)).convert("RGB")
init_image.thumbnail((768, 768))

prompt = "A fantasy landscape, trending on artstation"

images = pipe(
    prompt=prompt,
    num_inference_steps=20,
    image=init_image,
    # strength=0.75,
    guidance_scale=7.5,
).images

display(images[0])
```

**Comment by nipunjindal** [comment]
Note: MPS results are fixed with this [PR](https://github.com/huggingface/diffusers/pull/2643).

**Comment by nipunjindal** [comment]
@patrickvonplaten gentle reminder in case you could also share your feedback?

File: src/diffusers/schedulers/scheduling_dpmsolver_sde.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_dpmsolver_sde.py:20]: Agree, think having `torchsde` as a soft dependency here makes sense! The package is not too large and easily installable with 
```
pip install torchsde
```

File: src/diffusers/schedulers/__init__.py [hunk 3]
```diff
@@ -27,6 +27,7 @@
     from .scheduling_ddpm import DDPMScheduler
     from .scheduling_deis_multistep import DEISMultistepScheduler
     from .scheduling_dpmsolver_multistep import DPMSolverMultistepScheduler
+    from .scheduling_dpmsolver_sde import DPMSolverSDEScheduler
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/__init__.py:30]: we will need this under a `is_torchsde_available()` similarily to: https://github.com/huggingface/diffusers/blob/86ecd4b795f865b5b615b8c54991c177bb3dbef5/src/diffusers/__init__.py#L149

**Review by patrickvonplaten — COMMENTED** [review]
PR looks good to me - think we'll just have to:
- add a test
- add docs
- create a `is_torchsde_available()` function here: https://github.com/huggingface/diffusers/blob/86ecd4b795f865b5b615b8c54991c177bb3dbef5/src/diffusers/utils/import_utils.py#L346

Sorry for answering so late! Let me know if you need more help :-)

**Comment by nipunjindal** [comment]
Thanks so much, @patrickvonplaten 🙇. I'll put in the required thing pretty soon.

**Comment by nipunjindal** [comment]
Hello @patrickvonplaten @sayakpaul,

Thanks for reviewing my PR! I have addressed all of the previous comments and made the necessary changes. I would appreciate any further feedback before merging. Looking forward to your inputs.

File: tests/schedulers/test_scheduler_dpm_sde.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+from .test_schedulers import SchedulerCommonTest
+
+
+@require_torchsde
+class DPMSolverSDESchedulerTest(SchedulerCommonTest):
+    scheduler_classes = (DPMSolverSDEScheduler,)
+    num_inference_steps = 10
+
+    def get_scheduler_config(self, **kwargs):
+        # ??
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_dpm_sde.py:16]: remove this? 😅

**Review by yiyixuxu — COMMENTED** [review]
Thanks!

**Comment by nipunjindal** [comment]
@yiyixuxu Thank you! Addressed your feedback.

File: src/diffusers/schedulers/scheduling_heun_discrete.py [hunk 5]
```diff
@@ -70,8 +70,6 @@ class HeunDiscreteScheduler(SchedulerMixin, ConfigMixin):
             `linear` or `scaled_linear`.
         trained_betas (`np.ndarray`, optional):
             option to pass an array of betas directly to the constructor to bypass `beta_start`, `beta_end` etc.
-            options to clip the variance used when adding noise to the denoised sample. Choose from `fixed_small`,
-            `fixed_small_log`, `fixed_large`, `fixed_large_log`, `learned` or `learned_range`.
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_heun_discrete.py:74]: Is this an unrelated change?

File: src/diffusers/utils/testing_utils.py [hunk 6]
```diff
@@ -216,6 +217,13 @@ def require_note_seq(test_case):
     return unittest.skipUnless(is_note_seq_available(), "test requires note_seq")(test_case)
 
 
+def require_torchsde(test_case):
```
Comment (sayakpaul) [inline src/diffusers/utils/testing_utils.py:220]: Neat!

**Review by sayakpaul — APPROVED** [review]
Excellent work!

File: src/diffusers/schedulers/scheduling_heun_discrete.py (hunk 5 above)
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_heun_discrete.py:74]: Yes, based on this [comment](https://github.com/huggingface/diffusers/pull/3020#discussion_r1162245750), it wasn't correct in Heun as well. I can split it out to a different PR, if that's better?

Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_heun_discrete.py:74]: I think the change is small enough to be accommodated here. @yiyixuxu WDYT?

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_heun_discrete.py:74]: agreed - no need to open a new PR :)

**Comment by akx** [comment]
Um, the original code by Katherine Crowson [is licensed under the MIT license](https://github.com/crowsonkb/k-diffusion/blob/master/LICENSE). Is it acceptable to just copy-paste it into an Apache-licensed product?
